File: py/tests/bench/encodersTest.py
import json
import pytest

from stencilaschema.comms.jsonRpc import Request
from stencilaschema.comms.JsonEncoder import JsonEncoder
from stencilaschema.comms.JsonGzipEncoder import JsonGzipEncoder
from stencilaschema.comms.CborEncoder import CborEncoder
from stencilaschema.comms.CborGzipEncoder import CborGzipEncoder

from fixtures import fixture

encoders = dict(
    json=JsonEncoder(),
    jsongz=JsonGzipEncoder(),
    cbor=CborEncoder(),
    cborgz=CborGzipEncoder()
)

def read_fixture(filename):
    with open(fixture('requests', filename), 'r') as file:
        return JsonEncoder().decode(file.read(), Request)

requests = dict(
    record=read_fixture('record.json'),
    tree=read_fixture('tree.json'),
    table=read_fixture('table.json'),
    text=read_fixture('text.json')
)

matrix = {}
for ename, encoder in encoders.items():
    for rname, request in requests.items():
        matrix[f'{ename}-{rname}'] = (encoder, request)

sizes = {}

def roundtrip(encoder, request):
    message = encoder.encode(request)
    encoder.decode(message, Request)
    return len(message)

@pytest.mark.parametrize('name', list(matrix.keys()))
def test(benchmark, name):
    encoder, request = matrix[name]
    sizes[name] = benchmark(roundtrip, encoder, request)

def test_print_sizes():
    print(sizes)

File: js/tests/unit/comms/encoders.test.ts
import 'jest'
import each from 'jest-each'

import JsonRpcRequest from '../../../src/comms/JsonRpcRequest'
import JsonRpcResponse from '../../../src/comms/JsonRpcResponse'

import CborEncoder from '../../../src/comms/CborEncoder'
import CborGzipEncoder from '../../../src/comms/CborGzipEncoder'
import JsonEncoder from '../../../src/comms/JsonEncoder'
import JsonGzipEncoder from '../../../src/comms/JsonGzipEncoder'

each([
  ['cbor', CborEncoder],
  ['cbor+gzip', CborGzipEncoder],
  ['json', JsonEncoder],
  ['json+gzip', JsonGzipEncoder]
]).describe('%p', (name: string, Encoder: any) => {
  const encoder = new Encoder()
  
  test('name', () => {
    expect(encoder.name()).toEqual(name)
  })

  test('roundtrip', () => {
    const request1 = new JsonRpcRequest('answer', ['life', 'universe', 'everything'])
    const request2 = encoder.decode(encoder.encode(request1), JsonRpcRequest)
    expect(request2).toEqual(request1)

    const response1 = new JsonRpcResponse(21, 42)
    const response2 = encoder.decode(encoder.encode(response1), JsonRpcResponse)
    expect(response2).toEqual(response1)
  })
})
